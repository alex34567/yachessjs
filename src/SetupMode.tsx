import { useState } from 'react'
import { State } from './logic/state'
import ChessBoard from './ChessBoard'
import SetupInfo from './SetupInfo'
import * as React from 'react'
import { Pos } from './logic/util'
import { EMPTY, Piece } from './logic/pieces'
import { ModeProps } from './App'

export default function SetupMode (props: ModeProps) {
  const [highlightedPos, setHighlightedPos] = useState<Pos>()
  const [selectedPiece, setSelectedPiece] = useState<Piece>()

  const makeMove = (state: State) => props.setDefaultState(state)
  function changeHighlight (pos?: Pos) {
    setHighlightedPos(pos)
    setSelectedPiece(undefined)
  }
  function selectPiece (piece?: Piece) {
    if (highlightedPos) {
      props.setDefaultState(props.defaultState.modify(newState => {
        newState.board = newState.board.set(highlightedPos, EMPTY)
      }))
    } else {
      setSelectedPiece(piece)
    }
    setHighlightedPos(undefined)
  }

  const setup = { setupPiece: selectedPiece }
  return <div className='App'>
    <ChessBoard changeHighlight={changeHighlight} highlightedPos={highlightedPos} state={props.defaultState} setup={setup} makeMove={makeMove}/>
    <SetupInfo changeState={makeMove} selectPiece={selectPiece} selectedPiece={selectedPiece} switchMode={props.switchMode} state={props.defaultState}/>
  </div>
}
