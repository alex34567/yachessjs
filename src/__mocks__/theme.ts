import DEFAULT_THEME from '../defaultTheme.json'
import assert from 'assert'
import { Theme } from '../theme'
import { useState } from 'react'

export class ThemeManager {
  readonly theme: Readonly<Theme>
  readonly className: string = 'TestTheme'

  readonly style = document.createElement('style')

  constructor (theme?: Theme) {
    if (!theme) {
      theme = DEFAULT_THEME
    }
    this.theme = { ...theme }
    assert(/^[0-9a-zA-Z#]+$/.test(this.theme.boardSquare.whiteColor), 'Attempted xss')
    assert(/^[0-9a-zA-Z#]+$/.test(this.theme.boardSquare.blackColor), 'Attempted xss')
    this.style.textContent = `
      .ChessBoardSquareWhite.${this.className} {
        background-color: ${this.theme.boardSquare.whiteColor};
      }

      .ChessBoardSquareBlack.${this.className} {
       background-color: ${this.theme.boardSquare.blackColor};
      }
    `
  }
}

export function useTheme (defaultTheme?: Theme): [ThemeManager, (theme: Theme) => void] {
  const [theme, setTheme] = useState(new ThemeManager(defaultTheme))

  return [theme, theme => {
    setTheme(new ThemeManager(theme))
  }]
}
