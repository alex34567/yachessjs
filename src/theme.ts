import { useEffect, useState } from 'react'
import DEFAULT_THEME from './defaultTheme.json'
import assert from 'assert'

export interface BoardSquareTheme {
  name: string,
  id: string,
  // Must be an css color
  blackColor: string,
  whiteColor: string,
}

export interface PieceTheme {
  prefix: string,
  name: string,
}

export interface Theme {
  piece: PieceTheme
  boardSquare: BoardSquareTheme,
}

let CLASSNAME_SUFFIX = 0

export class ThemeManager {
  readonly theme: Readonly<Theme>
  readonly className: string

  readonly style = document.createElement('style')

  constructor (theme?: Theme) {
    if (!theme) {
      theme = DEFAULT_THEME
    }
    this.theme = { ...theme }
    this.className = `ChessTheme-${CLASSNAME_SUFFIX}`
    CLASSNAME_SUFFIX++
    assert(/^[0-9a-zA-Z#]+$/.test(this.theme.boardSquare.whiteColor), 'Attempted xss')
    assert(/^[0-9a-zA-Z#]+$/.test(this.theme.boardSquare.blackColor), 'Attempted xss')
    this.style.textContent = `
      .ChessBoardSquareWhite.${this.className} {
        background-color: ${this.theme.boardSquare.whiteColor};
      }

      .ChessBoardSquareBlack.${this.className} {
       background-color: ${this.theme.boardSquare.blackColor};
      }
    `
  }
}

export function useTheme (defaultTheme?: Theme): [ThemeManager, (theme: Theme) => void] {
  const [theme, setTheme] = useState(new ThemeManager(defaultTheme))
  useEffect(() => {
    document.head.append(theme.style)
    return () => {
      theme.style.remove()
    }
  }, [theme])

  return [theme, theme => {
    setTheme(new ThemeManager(theme))
  }]
}
