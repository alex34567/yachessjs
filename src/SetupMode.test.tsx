import { render, unmountComponentAtNode } from 'react-dom'
import { act } from '@testing-library/react'
import React from 'react'

import { ThemeManager } from './theme'
import SetupMode from './SetupMode'
import { BrowserRouter } from 'react-router-dom'

jest.mock('./BoardSquare')
jest.mock('./theme')

let container: HTMLDivElement | null = null
beforeEach(() => {
  // setup a DOM element as a render target
  container = document.createElement('div')
  document.body.appendChild(container)
})

afterEach(() => {
  // cleanup on exiting
  unmountComponentAtNode(container!)
  container!.remove()
  container = null
})

test('Setup Mode', () => {
  const dummy = jest.fn()

  act(() => {
    render(
      <BrowserRouter>
        <SetupMode openTheme={dummy} theme={new ThemeManager()}/>
      </BrowserRouter>
      , container)
  })

  expect(container).toMatchSnapshot()
})
