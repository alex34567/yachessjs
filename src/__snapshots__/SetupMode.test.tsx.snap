// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Setup Mode 1`] = `
<div>
  <div
    class="App"
  >
    <div
      class="PlayChessBoardBox"
    >
      <div
        class="ChessBoardSizer"
      >
        <div
          class="ChessBoard"
          style="height: 0px; width: 0px;"
        >
          <div
            data-name="r"
          />
          <div
            data-isblack="true"
            data-name="n"
          />
          <div
            data-name="b"
          />
          <div
            data-isblack="true"
            data-name="q"
          />
          <div
            data-name="k"
          />
          <div
            data-isblack="true"
            data-name="b"
          />
          <div
            data-name="n"
          />
          <div
            data-isblack="true"
            data-name="r"
          />
          <div
            data-isblack="true"
            data-name="p"
          />
          <div
            data-name="p"
          />
          <div
            data-isblack="true"
            data-name="p"
          />
          <div
            data-name="p"
          />
          <div
            data-isblack="true"
            data-name="p"
          />
          <div
            data-name="p"
          />
          <div
            data-isblack="true"
            data-name="p"
          />
          <div
            data-name="p"
          />
          <div />
          <div
            data-isblack="true"
          />
          <div />
          <div
            data-isblack="true"
          />
          <div />
          <div
            data-isblack="true"
          />
          <div />
          <div
            data-isblack="true"
          />
          <div
            data-isblack="true"
          />
          <div />
          <div
            data-isblack="true"
          />
          <div />
          <div
            data-isblack="true"
          />
          <div />
          <div
            data-isblack="true"
          />
          <div />
          <div />
          <div
            data-isblack="true"
          />
          <div />
          <div
            data-isblack="true"
          />
          <div />
          <div
            data-isblack="true"
          />
          <div />
          <div
            data-isblack="true"
          />
          <div
            data-isblack="true"
          />
          <div />
          <div
            data-isblack="true"
          />
          <div />
          <div
            data-isblack="true"
          />
          <div />
          <div
            data-isblack="true"
          />
          <div />
          <div
            data-name="P"
          />
          <div
            data-isblack="true"
            data-name="P"
          />
          <div
            data-name="P"
          />
          <div
            data-isblack="true"
            data-name="P"
          />
          <div
            data-name="P"
          />
          <div
            data-isblack="true"
            data-name="P"
          />
          <div
            data-name="P"
          />
          <div
            data-isblack="true"
            data-name="P"
          />
          <div
            data-isblack="true"
            data-name="R"
          />
          <div
            data-name="N"
          />
          <div
            data-isblack="true"
            data-name="B"
          />
          <div
            data-name="Q"
          />
          <div
            data-isblack="true"
            data-name="K"
          />
          <div
            data-name="B"
          />
          <div
            data-isblack="true"
            data-name="N"
          />
          <div
            data-name="R"
          />
        </div>
      </div>
    </div>
    <div
      class="SetupInfo"
    >
      <div
        class="SetupErrorBox"
      />
      <div
        class="SetupPieceBox"
      >
        <div
          data-name="p"
        />
        <div
          data-name="n"
        />
        <div
          data-name="b"
        />
        <div
          data-name="r"
        />
        <div
          data-name="q"
        />
        <div
          data-name="k"
        />
      </div>
      <div
        class="SetupPieceBox"
      >
        <div
          data-name="P"
        />
        <div
          data-name="N"
        />
        <div
          data-name="B"
        />
        <div
          data-name="R"
        />
        <div
          data-name="Q"
        />
        <div
          data-name="K"
        />
      </div>
      <div
        class="LabelledTextbox"
      >
        <label>
          Current Fen:
        </label>
        <input
          class="LabelledTextboxTextbox"
          readonly=""
          type="text"
          value="rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1"
        />
      </div>
      <div
        class="LabelledTextbox"
      >
        <label>
          Input Fen:
        </label>
        <input
          class="LabelledTextboxTextbox"
          value="rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1"
        />
      </div>
      <div>
        <label>
          First Move: 
        </label>
        <select>
          <option
            value="white"
          >
            White
          </option>
          <option
            value="black"
          >
            Black
          </option>
        </select>
      </div>
      <div>
        <label>
          En-passant file: 
        </label>
        <select>
          <option
            value=""
          >
            -
          </option>
          <option
            value="0"
          >
            a
          </option>
          <option
            value="1"
          >
            b
          </option>
          <option
            value="2"
          >
            c
          </option>
          <option
            value="3"
          >
            d
          </option>
          <option
            value="4"
          >
            e
          </option>
          <option
            value="5"
          >
            f
          </option>
          <option
            value="6"
          >
            g
          </option>
          <option
            value="7"
          >
            h
          </option>
        </select>
      </div>
      <div>
        <label>
          White O-O: 
        </label>
        <input
          checked=""
          type="checkbox"
        />
      </div>
      <div>
        <label>
          White O-O-O: 
        </label>
        <input
          checked=""
          type="checkbox"
        />
      </div>
      <div>
        <label>
          Black O-O: 
        </label>
        <input
          checked=""
          type="checkbox"
        />
      </div>
      <div>
        <label>
          Black O-O-O: 
        </label>
        <input
          checked=""
          type="checkbox"
        />
      </div>
      <div>
        <button>
          Import Fen
        </button>
        <button>
          Reset
        </button>
        <button>
          Clear
        </button>
      </div>
      <button>
        Play
      </button>
      <button>
        Change Theme
      </button>
    </div>
  </div>
</div>
`;
